### ISSUE TITLE
Guidance for integration of Typescript and Vuex 4 / Pinia 2

### ISSUE THREAD

**Issue by tohagan** [quasarframework/quasar#11222]
**Is your feature request related to a problem? Please describe.**

I'm needing some guidance  on how to use Typescript with Vuex v4 and Quasar 2.
Vuex v4 recommends registering Typescript typed Vuex stores using this code ...

- https://next.vuex.vuejs.org/guide/typescript-support.html#typing-usestore-composition-function

However Quasar 2 appears to registers the root store for you and unlike Quasar 1.x there's no `main.ts` or main.js anymore to apply `app.use()` before the app is mounted. I suspect that registering in a boot file is too late as the root store has already been created for you.  

```ts
// main.ts
import { createApp } from 'vue'
import { store, key } from './store'

const app = createApp({ ... })

// pass the injection key
app.use(store, key)  // <<< Probably too late to do this in a boot script.  Unclear if this is required *per store* or root store only??

app.mount('#app')
```

**Describe the solution you'd like**

Perhaps an update on the Quasar 2.x Vuex page with advice for Vuex 4 and Typescript. 

**Describe alternatives you've considered**

I also tried to use `pinia` with a boot script but hit other issues with Quasar 2. The `pinia` v2 stores lost my getters and actions for some reason - possibly related to conflicts with existing vuex store registration done by Quasar??. I ended up tossing the `pinia` stores and trying Vuex 4 only to hit another problem. I noted `pinia` had special instructions for Nuxt so probably Quasar needs some custom code to work with `pinia` v2 perhaps done in collaboration with the `pinia` team. 

- https://github.com/posva/pinia

**Comment by yusufkandemir** [comment]
Regarding Vuex, the injection key is already handled by Quasar for you. The documentation mentions the default store scaffolding and recommends using the `useStore` helper from `src/store`, see: https://quasar.dev/quasar-cli/vuex-store#typescript-support
Here is the default scaffolding, notice `storeKey` and `useStore`:
https://github.com/quasarframework/quasar-starter-kit/blob/master/template/src/store/index.ts
But Vuex 4 isn't doing a good job, especially with getters, with TS even if you use an injection key. There have been lots of mentions from the community about it on our Discord server. So, using Pinia is a better idea, especially for TS, atm.

For Pinia, you can use the regular store integration to create and return Pinia. Then define your stores and export them, then import them wherever you want, just regular Pinia stuff. Here is an example repo:
https://github.com/TobyMosque/quasar-v2-ssr-pinia
It might be better if we can add some mention/instructions to set it up and/or consider adding scaffolding/better integration for Pinia. But, as you can guess this needs extra effort, especially since we can't just ignore Vuex. So, any contribution is welcome as long as the solutions are not making things a lot more complex.

**Comment by tohagan** [comment]
Thanks very much ... It would very useful to add a link to Pina and this Pinia sample repo on the Quasar Vuex page.  Would have saved me 1/2 day of code refactoring multiple stores into Vuex. You could possibly rename the page to be about "Stores" rather than just "Vuex". I even had a crack and doing my own composition store uses `toRefs()` and `computed()` along the way.  

The secret sauce I was missing for integrating Pina 2 and Quasar 2 is here ... 
- https://github.com/TobyMosque/quasar-v2-ssr-pinia/tree/master/src/store
- https://github.com/TobyMosque/quasar-v2-ssr-pinia/blob/master/src/pages/Index.vue

I read that Vuex 4 getters are not yet caching (will be fixed soon).

**Comment by tohagan** [comment]
The example I needed for Typescript and Vuex 4 I found in a template for `store/index.ts`  inside Quasar ...

- https://github.com/quasarframework/quasar/blob/dev/app/templates/store/ts/index.ts

**Comment by nicojmb** [comment]
Hi, i successfuly integrated pinia in quasar 2!!!

in quasar.conf.js add pinia boot file
```
boot: [
      'i18n',
      'pinia'  // add this
    ],
```

in /src/boot, create a pinia.ts
```
import { boot } from 'quasar/wrappers'
import { createPinia } from 'pinia'

export default boot(({ app }) => {
  app.use(createPinia())
})
```
create a store file: /src/stores/user.ts
```
import { defineStore } from 'pinia'

const storeID = 'user'

interface IUser {
  loggedIn: boolean,
}

const User: IUser = {
  loggedIn: false
}

export { IUser }

export const useUserStore = defineStore({

  id: storeID,

  state: (): IUser => JSON.parse(localStorage.getItem(storeID) ?? JSON.stringify(User)) as IUser,

  actions: {
    login (username: string, password: string, remember = true): boolean {
      if (username === 'admin' && password === 'admin') {
        this.loggedIn = true
      }

      return this.loggedIn
    },
    logout (): boolean {
      this.loggedIn = false
      return true
    }
  },
  getters: {
    isLoggedIn (): boolean {
      return this.loggedIn
    }
  }
})

export type UserStore = ReturnType<typeof useUserStore>
```

use the store in any vue component/page/layout: 

```
<template>
Is logged in: {{ isLoggedIn  }}
</template>
<script lang="ts">
import { defineComponent, reactive } from 'vue'
import { useQuasar } from 'quasar'
import { useUserStore } from 'src/stores/user'

export default defineComponent({
  name: 'Login',
  setup () {

    const userStore = useUserStore()

    
    // example call a action
    userStore.login(form.username, form.password);

   // example call a getter
   const  isLoggedIn = user.isLoggedIn; ​

   ​return { 
      ​isLoggedIn 
   ​}
 ​}
})
</script>
```

and you can also use in routes to protect with router guards, example /src/router/index.ts

```
import { route } from 'quasar/wrappers'
import {
  createMemoryHistory,
  createRouter,
  createWebHashHistory,
  createWebHistory
} from 'vue-router'
import routes from './routes'
import { useUserStore } from 'src/stores/user'
/*
 * If not building with SSR mode, you can
 * directly export the Router instantiation;
 *
 * The function below can be async too; either use
 * async/await or return a Promise which resolves
 * with the Router instance.
 */

export default route(function (/* { store, ssrContext } */) {
  const createHistory = process.env.SERVER
    ? createMemoryHistory
    : (process.env.VUE_ROUTER_MODE === 'history' ? createWebHistory : createWebHashHistory)

  const Router = createRouter({
    scrollBehavior: () => ({ left: 0, top: 0 }),
    routes,

    // Leave this as is and make changes in quasar.conf.js instead!
    // quasar.conf.js -> build -> vueRouterMode
    // quasar.conf.js -> build -> publicPath
    history: createHistory(
      process.env.MODE === 'ssr' ? void 0 : process.env.VUE_ROUTER_BASE
    )
  })

  Router.beforeEach((to, from, next) => {
    if (to.matched.some(route => route.meta.requiresAuth)) {
      const user = useUserStore()

      if (!user.isLoggedIn) {
        next({ name: 'Login' })
      } else {
        next()
      }
    } else {
      next()
    }
  })

  return Router
})

```

and /src/router/routes.ts

```
import { RouteRecordRaw } from 'vue-router'

const routes: RouteRecordRaw[] = [
  {
    path: '/',
    component: () => import('layouts/MainLayout.vue'),
    children: [{ path: '', name: 'Dashboard', component: () => import('src/pages/Dashboard.vue') }],
    meta: {
      requiresAuth: true
    }
  },
  {
    path: '/auth',
    component: () => import('layouts/AuthLayout.vue'),
    children: [{ path: 'login', name: 'Login', component: () => import('pages/auth/Login.vue') }]
  },
  {
    path: '/:catchAll(.*)*',
    component: () => import('pages/errors/404.vue')
  }
]

export default routes
```

**Comment by yusufkandemir** [comment]
First-party Pinia support is released in [@quasar/app-vite-v1.0.0-beta.1](https://github.com/quasarframework/quasar/releases/tag/%40quasar%2Fapp-vite-v1.0.0-beta.1) and [@quasar/app-webpack-v3.4.0](https://github.com/quasarframework/quasar/releases/tag/%40quasar%2Fapp-v3.4.0). It's been available for almost a month now, so I am closing this issue. Thanks for the feature request!
